text/0000-macros-1.1.md

```diff
@@ -0,0 +1,552 @@
+[host-restriction]: https://github.com/rust-lang/rust/blob/78d49bfac2bbcd48de522199212a1209f498e834/src/librustc_metadata/creader.rs#L494
+
+Once a crate is found, it's required to be a dynamic library as well, and once
+that's all verified the compiler [opens it up with `dlopen`][dlopen] (or the
+equivalent therein). After loading, the compiler will [look for a special
+symbol][symbol] in the dynamic library, and then call it with a macro context.
+
+[dlopen]: https://github.com/rust-lang/rust/blob/78d49bfac2bbcd48de522199212a1209f498e834/src/librustc_plugin/load.rs#L124
+[symbol]: https://github.com/rust-lang/rust/blob/78d49bfac2bbcd48de522199212a1209f498e834/src/librustc_plugin/load.rs#L136-L139
+
+So as we've seen macros are compiled as normal crates into dynamic libraries.
+One function in the crate is tagged with `#[plugin_registrar]` which gets wired
+up to this "special symbol" the compiler wants. When the function is called with
+a macro context, it uses the passed in [plugin registry][registry] to register
+custom macros, attributes, etc.
+
+[registry]: https://github.com/rust-lang/rust/blob/78d49bfac2bbcd48de522199212a1209f498e834/src/librustc_plugin/registry.rs#L30-L69
+
+After a macro is registered, the compiler will then continue the normal process
+of expanding a crate. Whenever the compiler encounters this macro it will call
+this registration with essentially and AST and morally gets back a different
+AST to splice in or replace.
+
+### Today's drawbacks
+
+This expansion process suffers from many of the downsides mentioned in the
+motivation section, such as a lack of hygiene, a lack of modularity, and the
+inability to import macros as you would normally other functionality in the
+module system.
+
+Additionally, though, it's essentially impossible to ever *stabilize* because
+the interface to the compiler is... the compiler! We clearly want to make
+changes to the compiler over time, so this isn't acceptable. To have a stable
+interface we'll need to cut down this surface area *dramatically* to a curated
+set of known-stable APIs.
+
+Somewhat more subtly, the technical ABI of procedural macros is also exposed
+quite thinly today as well. The implementation detail of dynamic libraries, and
+especially that both the compiler and the macro dynamically link to libraries
+like libsyntax, cannot be changed. This precludes, for example, a completely
+statically linked compiler (e.g. compiled for `x86_64-unknown-linux-musl`).
+Another goal of this RFC will also be to hide as many of these technical
+details as possible, allowing the compiler to flexibly change how it interfaces
+to macros.
+
+## Macros 1.1
+
+Ok, with the background knowledge of what procedural macros are today, let's
+take a look at how we can solve the major problems blocking its stabilization:
+
+* Sharing an API of the entire compiler
+* Frozen interface between the compiler and macros
+
+### `librustc_macro`
```

Comment:
This is currently done because `macro` is a keyword, and this doesn't propose changing that just yet. I was thinking that we could perhaps wholesale rename the crate one day as well if `macro` becomes an available name.